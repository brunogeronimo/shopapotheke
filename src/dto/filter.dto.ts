import { IsArray, IsDate, IsDateString, IsEnum, IsIn, IsISO8601, IsOptional, IsString } from 'class-validator';
import { Limit } from '../enums/limit.enum';
import { Sort } from '../enums/sort.enum';
import { ApiProperty } from '@nestjs/swagger';

export class FilterDto {
  @ApiProperty({
    name: 'languages[]',
    required: false,
    isArray: true,
    type: 'string',
    description: 'Filter by programming language'
  })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  languages: string[];

  @ApiProperty({ required: false, default: '10' })
  @IsOptional()
  @IsEnum(Limit)
  limit: Limit;

  @ApiProperty({ required: false, default: Sort.STARS_DESC })
  @IsOptional()
  @IsEnum(Sort)
  sortBy: Sort;

  @ApiProperty({ required: false })
  @IsOptional()
  @IsString()
  text: string;

  @ApiProperty({ required: false, description: 'Format: YYYY-MM-DD' })
  @IsOptional()
  @IsISO8601()
  createdAfter: string;

  @ApiProperty({ required: false, description: 'Format: YYYY-MM-DD' })
  @IsOptional()
  @IsISO8601()
  createdBefore: string;
}