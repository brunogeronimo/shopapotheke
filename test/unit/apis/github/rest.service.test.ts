import { RestService } from '../../../../src/apis/github/rest.service';
import { EMPTY_RESPONSE, VALID_RESPONSE } from './assets/rest.respose.mock';
import { FilterDto } from '../../../../src/dto/filter.dto';
import axios from "axios";
import { Limit } from '../../../../src/enums/limit.enum';
import { Sort } from '../../../../src/enums/sort.enum';
import { DateFilter } from '../../../../src/apis/github/enums/date-filter.enum';

jest.mock('axios');

describe('REST Service', () => {
  let restService: RestService;

  beforeEach(() => {
    jest.clearAllMocks();

    restService = new RestService();
  });

  it('Should parse response properly', async () => {
    // @ts-ignore
    axios.get.mockResolvedValue({
      data: VALID_RESPONSE
    });

    const repositories = await restService.search(new FilterDto());

    expect(axios.get).toBeCalledWith(
      expect.any(String),
      expect.objectContaining({params: {
          per_page: "10",
          q: "sort:stars-desc"
      }})
    );
    expect(repositories).toMatchSnapshot();
  });

  it('Should not break if there are no repositories in the response', async () => {
    // @ts-ignore
    axios.get.mockResolvedValue({
      data: EMPTY_RESPONSE
    });

    const repositories = await restService.search(new FilterDto());

    expect(axios.get).toBeCalled();
    expect(repositories).toMatchObject([]);
  });

  it('Should dispatch the error if something goes wrong', async (done) => {
    // @ts-ignore
    jest.spyOn(console, 'error').mockImplementation();

    const rejectionMock = jest.fn();

    // @ts-ignore
    axios.get.mockImplementationOnce(() => {
      return new Promise((resolved, rejected) => {
        rejected(rejectionMock);
      });
    });

    restService.search(new FilterDto()).catch(reject => {
      expect(axios.get).toBeCalled();
      expect(console.error).toBeCalled();
      done();
    });
  });

  it.each(
    [
      [
        {
          languages: ['PHP', 'Clojure']
        },
        'lang:PHP Clojure'
      ],
      [
        {
          sortBy: Sort.STARS_DESC
        },
        'sort:stars-desc'
      ],
      [
        {
          limit: Limit.ONE_HUNDRED
        },
        '100'
      ],
      [
        {
          text: 'foo text'
        },
        'foo text'
      ],
      [
        {
          createdAfter: '2012-12-12'
        },
        `${DateFilter.AFTER}2012-12-12`
      ],
      [
        {
          createdBefore: '2020-01-01'
        },
        `${DateFilter.BEFORE}2020-01-01`
      ]
    ]
  )('Adds filters to query', async (filter: Object, searchString: string) => {
    // @ts-ignore
    axios.get.mockResolvedValue({
      data: EMPTY_RESPONSE
    });

    const filterDto = new FilterDto();
    const key = Object.keys(filter)[0];

    filterDto[key] = filter[key];

    await restService.search(filterDto);

    let validation;

    if (filterDto.limit) {
      validation = { per_page: expect.stringContaining(searchString) }
    } else {
      validation = { q: expect.stringContaining(searchString) }
    }

    expect(axios.get).toBeCalledWith(
      expect.any(String),
      expect.objectContaining({
        params: expect.objectContaining(validation)
      })
    );
  })

});