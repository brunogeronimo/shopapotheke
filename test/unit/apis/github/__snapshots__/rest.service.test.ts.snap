// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REST Service Should parse response properly 1`] = `
Array [
  Repository {
    "createdAt": 2011-11-29T22:07:23.000Z,
    "description": "A collection of JUnit rules for testing code which uses java.lang.System.",
    "name": "system-rules",
    "owner": "stefanbirkner",
    "stars": 462,
    "url": "https://api.github.com/repos/stefanbirkner/system-rules",
  },
  Repository {
    "createdAt": 2016-09-23T09:34:46.000Z,
    "description": "‰æùËµñ‰∫é org.common.lang ÁöÑ java Â∑•ÂÖ∑Â∫ì",
    "name": "tool",
    "owner": "OrsonEx",
    "stars": 77,
    "url": "https://api.github.com/repos/OrsonEx/tool",
  },
  Repository {
    "createdAt": 2014-11-18T08:12:22.000Z,
    "description": "Better / faster proxy generator than java.lang.reflect.Proxy for Java (require 1.7)",
    "name": "proxy2",
    "owner": "forax",
    "stars": 55,
    "url": "https://api.github.com/repos/forax/proxy2",
  },
  Repository {
    "createdAt": 2015-01-16T11:58:20.000Z,
    "description": "Utilize Java Agents and ASM to modify then override Java SE classes. (In this case blocking java.lang.Runtime.exec)",
    "name": "JVM-Sandbox",
    "owner": "Konloch",
    "stars": 35,
    "url": "https://api.github.com/repos/Konloch/JVM-Sandbox",
  },
  Repository {
    "createdAt": 2015-05-04T11:33:14.000Z,
    "description": "üçã FFI example for accessing Rust lang dynamic libraries from java ",
    "name": "java-rust-ffi",
    "owner": "sureshg",
    "stars": 30,
    "url": "https://api.github.com/repos/sureshg/java-rust-ffi",
  },
  Repository {
    "createdAt": 2016-03-08T20:36:00.000Z,
    "description": "System Lambda is a collection of functions for testing code that uses java.lang.System",
    "name": "system-lambda",
    "owner": "stefanbirkner",
    "stars": 30,
    "url": "https://api.github.com/repos/stefanbirkner/system-lambda",
  },
  Repository {
    "createdAt": 2014-02-13T04:07:22.000Z,
    "description": "Pojomatic provides configurable implementations of the equals(Object), hashCode() and toString() methods inherited from java.lang.Object.",
    "name": "pojomatic",
    "owner": "irobertson",
    "stars": 27,
    "url": "https://api.github.com/repos/irobertson/pojomatic",
  },
  Repository {
    "createdAt": 2016-05-26T02:25:33.000Z,
    "description": "ÂÆû‰æãËØ¶Ëß£ java.lang.OutOfMemoryError",
    "name": "outofmemoryerror",
    "owner": "cncounter",
    "stars": 25,
    "url": "https://api.github.com/repos/cncounter/outofmemoryerror",
  },
  Repository {
    "createdAt": 2018-05-28T13:47:42.000Z,
    "description": "Experiments for \\"Enough java.lang.String to Hang Ourselves ...\\"",
    "name": "string-performance",
    "owner": "kabutz",
    "stars": 22,
    "url": "https://api.github.com/repos/kabutz/string-performance",
  },
  Repository {
    "createdAt": 2018-08-10T06:40:05.000Z,
    "description": "TypeScript Enum like java.lang.Enum",
    "name": "ts-jenum",
    "owner": "reforms",
    "stars": 20,
    "url": "https://api.github.com/repos/reforms/ts-jenum",
  },
]
`;
